### PR TITLE
Downlink parser software

### PR DESCRIPTION
**Pull request by tanishqaggarwal**

### CODE REVIEW

File: platformio.ini [hunk 1]
```diff
@@ -57,7 +58,7 @@ platform = native
 lib_deps =
   https://github.com/pathfinder-for-autonomous-navigation/psim
   https://github.com/pathfinder-for-autonomous-navigation/CommonSoftware
-  ;file:///Users/athena/PAN_2019/CommonSoftware
+  ;file:///Users/tanishqaggarwal/Documents/pan/repositories/CommonSoftware
```
Comment (tanishqaggarwal) [inline platformio.ini]: Sorry @athena255

File: src/FCCode/MissionManager.cpp [hunk 2]
```diff
@@ -25,8 +25,9 @@ MissionManager::MissionManager(StateFieldRegistry& registry, unsigned int offset
     assert(adcs_ang_rate_fp);
     assert(adcs_min_stable_ang_rate_fp);
 
-    // TODO change to startup.
-    mission_mode_f.set(static_cast<unsigned int>(mission_mode_t::detumble));
+    mission_mode_f.set(static_cast<unsigned int>(mission_mode_t::startup));
+    is_deployed_f.set(false);
+    sat_designation_f.set(0);
```
Comment (tanishqaggarwal) [inline src/FCCode/MissionManager.cpp]: Drive-by initialization fixes.

File: test/test_dummy_fsw.py [hunk 3]
```diff
@@ -24,6 +24,8 @@ def setUp(self):
         self.console = serial.Serial(os.ttyname(slave_fd), 9600, timeout=1)
```
Comment (tanishqaggarwal) [inline test/test_dummy_fsw.py]: This file just contains drive-by documentation fixes.

File: test/test_real_fsw.py [hunk 4]
```diff
@@ -24,6 +24,8 @@ def setUp(self):
         self.console = serial.Serial(os.ttyname(slave_fd), 9600, timeout=1)
```
Comment (tanishqaggarwal) [inline test/test_real_fsw.py]: Drive-by documentation fixes.

File: test/test_gsw_downlink_parser/test_gsw_downlink_parser.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            "foo1"
+        }
+    }
+};
+
+void test_task_initialization() {
+    TestFixture tf;
+}
+
+void test_task_execute() {
```
Comment (tanishqaggarwal) [inline test/test_gsw_downlink_parser/test_gsw_downlink_parser.cpp]: This runs one execution of Flight Software and processes the packet produced by flight software.

File: src/GroundCode/DownlinkParser.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        ret["metadata"]["flow_ids"] = json::array();
+
+        // Process the downlink frame in four steps.
+
+        // Step 1: Manage the downlink frame at a bit level.
+        std::vector<bool> frame_bits(frame_to_process.size() * 8);
+        for(size_t i = 0; i < frame_to_process.size(); i++) {
+            std::bitset<8> x(frame_to_process[i]);
+            for(int j = 0; j < 8; j++) {
+                frame_bits[i*8 + j] = x[7 - j];
```
Comment (athena255) [inline src/GroundCode/DownlinkParser.cpp]: won't this swap bit order, as in MSB is placed in an earlier index than LSB

File: src/GroundCode/DownlinkParser.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            // to the downlink data.
+            for(ReadableStateFieldBase* field : flow->field_list) {
+                const std::vector<bool>::iterator field_end_it =
+                    frame_bits.begin() + field->get_bit_array().size();
+
+                const std::vector<bool> field_bits(frame_bits.begin(), field_end_it);
+                field->set_bit_array(field_bits);
+                field->deserialize();
+                ret["data"][field->name()] = std::string(field->print());
+                frame_bits.erase(frame_bits.begin(), field_end_it);
```
Comment (athena255) [inline src/GroundCode/DownlinkParser.cpp]: Depending on how many fields in field_list, might be more efficient to declare variables (field_bits, field_end_it) outside the for loop.

File: src/GroundCode/DownlinkParser.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        // size of a field exceeds the number of bits available in the
+        // downlink, then stop processing.
+        //
+        // As flows are processed, their bits are removed from the frame.
+        while(frame_bits.size() > 0) {
+            // Step 4.1. Get flow ID and check if it's valid.
+            unsigned char flow_id;
+
+            Serializer<unsigned char> flow_id_sr(flow_data.size());
+            if (flow_id_sr.bitsize() > frame_bits.size()) {
```
Comment (athena255) [inline src/GroundCode/DownlinkParser.cpp]: Should this be >= ?

File: src/GroundCode/DownlinkParser.cpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+std::string DownlinkParser::process_downlink_packet(const std::vector<char>& packet) {
+    // The returned JSON object.
+    using json = nlohmann::json;
+    json ret;
+
+    // If the first bit of the packet is a 1, it's the start of a new frame.
+    // Otherwise the packet belongs to a previous frame.
+    const bool is_first_packet_in_frame = (static_cast<unsigned char>(packet[0]) >> 7) == 0b1;
```
Comment (athena255) [inline src/GroundCode/DownlinkParser.cpp]: This gets the most significant bit. Is that intended?

File: src/GroundCode/DownlinkParser.cpp [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        std::vector<bool> frame_bits(frame_to_process.size() * 8);
+        for(size_t i = 0; i < frame_to_process.size(); i++) {
+            std::bitset<8> x(frame_to_process[i]);
+            for(int j = 0; j < 8; j++) {
+                frame_bits[i*8 + j] = x[7 - j];
+            }
+        }
+
+        // Step 2: Remove header bits from the packet.
+        for(int i = frame_bits.size() / 560; i >= 0; i--) {
```
Comment (athena255) [inline src/GroundCode/DownlinkParser.cpp]: What is 560? Max bits in a frame?

File: src/GroundCode/DownlinkParser.cpp (hunk 9 above)
Comment (tanishqaggarwal) [reply src/GroundCode/DownlinkParser.cpp]: The most significant bit is the first bit in the packet, right?

File: src/GroundCode/DownlinkParser.cpp (hunk 6 above)
Comment (tanishqaggarwal) [reply src/GroundCode/DownlinkParser.cpp]: I think that's the bit order inside the downlink. The tests pass and they definitely won't pass the other way around.

File: src/GroundCode/DownlinkParser.cpp (hunk 10 above)
Comment (tanishqaggarwal) [reply src/GroundCode/DownlinkParser.cpp]: Yep

File: src/GroundCode/DownlinkParser.cpp (hunk 8 above)
Comment (tanishqaggarwal) [reply src/GroundCode/DownlinkParser.cpp]: Yeah I agree, thanks for catching

File: src/GroundCode/DownlinkParser.cpp (hunk 7 above)
Comment (tanishqaggarwal) [reply src/GroundCode/DownlinkParser.cpp]: It might be more efficient, but I think it's marginally more readable to declare variables as they get used. In any case efficiency doesn't matter too much given that there are < 200 readable state fields, and this code must run on a desktop platform

File: src/GroundCode/DownlinkParser.cpp (hunk 8 above)
Comment (tanishqaggarwal) [reply src/GroundCode/DownlinkParser.cpp]: Actually nevermind, I think it should be `>`. If the bitsize is equal to the number of remaining bits in the frame, then the flow ID will be completely contained within the frame. It's a separate matter that there might not be any data in the flow; this is caught by later assertions.
